base/applications/rapps/gui.cpp

```diff
@@ -541,37 +649,87 @@ class CAppSnapshotPreview :
             delete pImage;
             pImage = NULL;
         }
+        if (AsyncInet)
+        {
+            AsyncInetCancel(AsyncInet);
+            AsyncInet = NULL;
+        }
+        if (!TempImagePath.IsEmpty())
+        {
+            DeleteFileW(TempImagePath.GetString());
+            TempImagePath.Empty();
+        }
     }
 
     VOID DisplayEmpty()
     {
+        InterlockedIncrement64(&ContentID);
         SetStatus(SNPSHTPREV_EMPTY);
         PreviousDisplayCleanup();
     }
 
-    VOID DisplayLoading()
+    BOOL DisplayImage(LPCWSTR lpszLocation)
     {
-        SetStatus(SNPSHTPREV_LOADING);
+        LONGLONG ID = InterlockedIncrement64(&ContentID);
         PreviousDisplayCleanup();
-        bLoadingTimerOn = TRUE;
-        SetTimer(TIMER_LOADING_ANIMATION, 1000 / LOADING_ANIMATION_FPS, 0);
-    }
 
-    BOOL DisplayFile(LPCWSTR lpszFileName)
-    {
-        SetStatus(SNPSHTPREV_FILE);
-        PreviousDisplayCleanup();
-        pImage = Bitmap::FromFile(lpszFileName, 0);
-        if (pImage->GetLastStatus() != Ok)
+        if (PathIsURLW(lpszLocation))
         {
-            DisplayFailed();
-            return FALSE;
+            DisplayLoading();
+
+            SNPSHT_DOWNLOAD_PARAM* DownloadParam = new SNPSHT_DOWNLOAD_PARAM;
+            if (!DownloadParam) return FALSE;
+
+            DownloadParam->hwndNotify = m_hWnd;
+            DownloadParam->ID = ID;
+            // generate a filename
+            ATL::CStringW SnapshotFolder = CAvailableApps::m_Strings.szAppsPath;
+            PathAppendW(SnapshotFolder.GetBuffer(MAX_PATH), L"snapshots");
+            SnapshotFolder.ReleaseBuffer();
+
+            if (!PathIsDirectoryW(SnapshotFolder.GetString()))
+            {
+                CreateDirectoryW(SnapshotFolder.GetString(), NULL);
+            }
+
+            if (!GetTempFileNameW(SnapshotFolder.GetString(), L"img",
+                0, DownloadParam->DownloadFileName.GetBuffer(MAX_PATH)))
+            {
+                DownloadParam->DownloadFileName.ReleaseBuffer();
+                delete DownloadParam;
+                DisplayFailed();
+                return FALSE;
+            }
+            DownloadParam->hFile = CreateFileW(DownloadParam->DownloadFileName.GetBuffer(),
```

Comment:
`GetString()` or completely omit the function, just passing `DownloadParam->DownloadFileName` (since the CStringW overloads an operator that will ensure this is casted properly)